doc/overview.md

```diff
@@ -0,0 +1,124 @@
+lists. A second and even more interesting use-case for a graph are the actions
+performed when committing the queued changes to the system.
+
+
+### Device Graph
+
+Storing the devices in libstorage as a graph looks like a natural
+approach. Here is a simple graph showing a disk with a partition table, two
+partitions and two filesystems where one is encrypted:
+
+![luks.png](images/luks.png)
+
+You might have seen the diagram above in YaST. But in libstorage-ng it is a
+direct dump of internal data structure instead of a [cumbersome
+transformation](https://github.com/openSUSE/libstorage/blob/master/storage/Graph.cc).
+
+The graph approach has several advantages compared to the list of list design
+of libstorage:
+
+* No pseudo containers for e.g. MD RAID, tmpfs or NFS are needed.
+
+* Possibility to have special objects for IMSM and DDF RAID containers.
+
+* Some operations are trivial compared to the current libstorage, e.g. getting
+  all devices using a device is just a plain
+  [BFS](https://en.wikipedia.org/wiki/Breadth-first_search).
+
+Together with a redesign of the device objects several features will be
+simple, e.g. using unpartitioned disks for filesystems, renaming LVM volume
+groups and logical volumes, switching between partitioned and unpartitioned MD
+RAID.
+
+
+### Action Graph
+
+When committing the queued changes to the system libstorage-ng has to analyse
+what steps are needed and in which order they must be done. Unfortunately the
+code in libstorage has evolved and many if statements span several lines of
+code.
+
+The new approach of libstorage-ng looks different:
+
+* Two device graphs (the first is the current system, the second the staging
+  system) are compared and the required steps are stored as vertices of a
+  graph. E.g. a step is to create a partition.
+
+* Dependencies are added as edges to the graph. E.g. when creating two
+  partitions on a disk the ordering is important (otherwise the partition
+  numbers get swapped).
+
+* A topological sort calculates the order in which the steps can be executed.
+
+Here is a simply action graph for creating two partitions used by a volume
+group with one logical volume.
+
+![action.png](images/action.png)
+
+Legend: The color indicates whether an action creates (green), modifies (blue)
+or deletes (red) and object.
+
+The new approach of libstorage-ng has several advantages:
+
+* Better testability: The code to generate the actions and their dependencies
+  can be checked in a testsuite. Only testing the actions themself needs to be
+  done on an actual system.
+
+* Checking the dependencies instead of the ordering is more robust. The
+  ordering can be correct "by luck" while dependencies are explicit.
+
+* The topological sort can also be used to find actions that can be done in
+  parallel. This might be used for filesystem creation (although newer
+  filesystems (e.g. btrfs, xfs, ext4) have fast mkfs commands).
```

Comment:
Nested parentheses look quite weird in English. You could remove the external ones.

For example, this might be used for filesystem creation, although newer filesystems (like btrfs, xfs or ext4) have fast mkfs commands.